core/src/main/java/org/apache/druid/java/util/common/ByteBufferUtils.java

```diff
@@ -19,16 +19,124 @@
+      UNMAP_NOT_SUPPORTED_EXCEPTION = null;
+    } else {
+      UNMAP = null;
+      UNMAP_NOT_SUPPORTED_EXCEPTION = exception;
+    }
+  }
+
+  private static void clean(ByteBuffer buffer)
+  {
+    if (!buffer.isDirect()) {
+      throw new IllegalArgumentException("Unmapping only works with direct buffers");
+    }
+    if (UNMAP == null) {
+      throw UNMAP_NOT_SUPPORTED_EXCEPTION;
+    }
+
+    try {
+      UNMAP.invokeExact(buffer);
+    }
+    catch (Throwable throwable) {
+      throw new RuntimeException("Unable to unmap the mapped buffer", throwable);
+    }
+  }
+
+  private static MethodHandle lookupUnmapMethodHandle()
+  {
+    final MethodHandles.Lookup lookup = MethodHandles.lookup();
+    try {
+      if (JvmUtils.isIsJava9Compatible()) {
+        return unmapJava9(lookup);
+      } else {
+        return unmapJava7Or8(lookup);
+      }
+    }
+    catch (ReflectiveOperationException | RuntimeException e1) {
+      throw new UnsupportedOperationException("Unmapping is not supported on this platform, because internal " +
+                                              "Java APIs are not compatible with this Druid version", e1);
+    }
+  }
+
+  private static MethodHandle unmapJava7Or8(MethodHandles.Lookup lookup) throws ReflectiveOperationException
+  {
+    // "Compile" a MethodHandle that is roughly equivalent to the following lambda:
+    //
+    // (ByteBuffer buffer) -> {
+    //   sun.misc.Cleaner cleaner = ((java.nio.DirectByteBuffer) byteBuffer).cleaner();
+    //   if (nonNull(cleaner))
+    //     cleaner.clean();
+    //   else
+    //     noop(cleaner); // the noop is needed because MethodHandles#guardWithTest always needs both if and else
+    // }
+    //
+    Class<?> directBufferClass = Class.forName("java.nio.DirectByteBuffer");
+    Method m = directBufferClass.getMethod("cleaner");
+    m.setAccessible(true);
+    MethodHandle directBufferCleanerMethod = lookup.unreflect(m);
+    Class<?> cleanerClass = directBufferCleanerMethod.type().returnType();
+    MethodHandle cleanMethod = lookup.findVirtual(cleanerClass, "clean", MethodType.methodType(void.class));
+    MethodHandle nonNullTest = lookup.findStatic(ByteBufferUtils.class, "nonNull",
+                                                 MethodType.methodType(boolean.class, Object.class)
+    ).asType(MethodType.methodType(boolean.class, cleanerClass));
+    MethodHandle noop = MethodHandles.dropArguments(MethodHandles.constant(
+        Void.class,
+        null
+    ).asType(MethodType.methodType(void.class)), 0, cleanerClass);
+    MethodHandle unmapper = MethodHandles.filterReturnValue(
+        directBufferCleanerMethod,
+        MethodHandles.guardWithTest(nonNullTest, cleanMethod, noop)
+    ).asType(MethodType.methodType(void.class, ByteBuffer.class));
+    return unmapper;
+  }
+
+  private static MethodHandle unmapJava9(MethodHandles.Lookup lookup) throws ReflectiveOperationException
+  {
+    Class<?> unsafeClass = Class.forName("sun.misc.Unsafe");
```

Comment:
This is done in a slightly different way in

https://github.com/apache/incubator-druid/blob/0e926e865276f892ddc5b62685d331b822104022/processing/src/main/java/org/apache/druid/query/monomorphicprocessing/SpecializationService.java#L68-L79 

can we unify these?

Reply:
That’s another one we need to address. It’s a different beast though, I’ll submit in another PR for that.